### ISSUE TITLE
Require Dot Notation Whenever Possible (dot-notation)

### ISSUE THREAD

**Issue by julien-f** [standard/standard#1344]
Hey folks, what do you think about enabling the [dot-notation rule](https://eslint.org/docs/rules/dot-notation)?

**Comment by feross** [comment]
Modest ecosystem impact:

```
1..162
# tests 162
# pass  153
# fail  9
```

And they're all cases where dot notation would be a bit better. Since the rule is automatically fixable, I think this rule is a go for `standard` v14.

**Comment by feross** [comment]
After fixing repos I control, only 3 ecosystem repos fail.

**Comment by LinusU** [comment]
`application-config-path` is mine, I can fix it 👍 

Funnily enough, it's exactly the case I described in my first comment and I actually think that the code will be a bit uglier 😬 

Oh well, it's for the greater good 🚀

**Comment by feross** [comment]
Yeah, I actually agree with you. There's a small amount information that it feels like we're losing by always using dot notation. I noticed when I went through all the failing cases that I was using `[]` consistently in very specific situations like HTTP headers: `res.headers['accept']`. It's hard to put into words why this way feels better than dot notation for these certain cases, but it does.

Happy to reconsider this if we end up not liking it. But it's only required changes in a few places so I'm guessing it won't be that big of a deal :)

**Comment by LinusU** [comment]
> [...] using the `[...]` syntax when I have an object that's being used as a map.

I think this is what does it for me. JavaScript doesn't really have a distinction between objects with specific properties, and just general maps/dictionaries (or, it has now `Map`, but many APIs came before that. You can see this clearly with TypeScript typing: `process.env` and `res.headers` are described as `Record<string, ...>` whereas objects have there specific keys described.

It would be awesome if it could somehow be enforced that `process.env`, `req.headers`, etc. always have `[...]` notation, but I think that's impossible 😄

**Comment by vweevers** [comment]
> It's hard to put into words why this way feels better than dot notation for these certain cases, but it does.

In the case of `headers`, it feels better because you may have (or later add) headers that do require `[...]` syntax. For visual consistency, I prefer:

```js
headers['user-agent'] = 'example'
headers['connection'] = 'keep-alive'
headers['accept-encoding'] = 'gzip, deflate, identity'
```

Over:

```js
headers['user-agent'] = 'example'
headers.connection = 'keep-alive'
headers['accept-encoding'] = 'gzip, deflate, identity'
```

**Comment by mightyiam** [comment]
Perhaps I would write that way, as well, @vweevers. If you could contribute that as an option to the ESLint rule, then we'd have the option to consider it. I personally feel that it would be better to choose a not-always pleasing style over not choosing.

**Comment by jimmywarting** [comment]
crap, now i must override this rule where i use [closure compiler](https://developers.google.com/closure/compiler/docs/api-tutorial3) in advance mode

didn't like this addition.
ppl use dot notation anyways? if not then they have a reason for it?

Getting more and more tiered of linters that is strict about formatting. hence why i don't use them so much any longer.
all i want is a linter that can spot errors, (telling me if i missed a variable declaration, if i reassigning const variables etc) not tell me how i should format my code 😞

**Comment by feross** [comment]
> @mightyiam  I personally feel that it would be better to choose a not-always pleasing style over not choosing.

I generally feel this way too.

> @jimmywarting ppl mostly use dot notation anyways? if not then they have a reason for it.

In general, this is probably true. The ecosystem impact of this change was low.

> now i must override this rule where i use closure compiler in advance mode

I didn't realize that Closure Compiler was still a popular compiler. Does anyone else use Closure and is affected by this change?

> all i want is a linter that can only spot errors

You could try doing using eslint-config-standard and eslint-config-prettier (which disables style-related rules)

**Comment by feross** [comment]
I'll re-open this issue for a little while to give it visibility. If others have issues with this rule, we can reconsider it for the next minor version.
